### PR TITLE
feature: make hco crd a singleton

### PR DESCRIPTION
**Pull request by djzager**
Update the HCO CustomResourceDefinition to accept only resources of a
particular name|namespace.

Signed-off-by: David Zager <dzager@redhat.com>

### CODE REVIEW

File: pkg/components/components.go [hunk 1]
```diff
@@ -426,6 +422,21 @@ func GetOperatorCRD() *extv1beta1.CustomResourceDefinition {
...
+				OpenAPIV3Schema: &extv1beta1.JSONSchemaProps{
+					Properties: map[string]extv1beta1.JSONSchemaProps{
+						"name": extv1beta1.JSONSchemaProps{
+							Type:    "string",
+							Pattern: hcov1alpha1.HyperConvergedName,
+						},
+						"namespace": extv1beta1.JSONSchemaProps{
+							Type:    "string",
+							Pattern: namespace,
+						},
```
Comment (orenc1) [inline pkg/components/components.go]: This validation hierarchy is not correct. It should be:
validation.openAPIV3Schema.properties.**metadata.properties**
see:
https://github.com/kubernetes-sigs/kubebuilder/issues/1074
I checked both yamls and the latter works, the former doesn't.

Comment (djzager) [reply pkg/components/components.go]: good catch

**Comment by djzager** [comment]
Part 1, restricting the namespaced name of the HyperConverged resource:

```
$ cat deploy/hco.cr.yaml
---
apiVersion: hco.kubevirt.io/v1alpha1
kind: HyperConverged
metadata:
  name: kubevirt-hyperconverged
  namespace: default

$ kubectl create -f deploy/hco.cr.yaml
The HyperConverged "kubevirt-hyperconverged" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"hco.kubevirt.io/v1alpha1", "kind":"HyperConverged", "metadata":map[string]interface {}{"creationTimestamp":"2020-03-04T17:40:21Z", "generation":1, "name":"kubevirt-hyperconverged", "namespace":"default", "uid":"387e4b54-5e3f-11ea-83f0-52fdfc072182"}}: validation failure list:
metadata.namespace in body should match 'kubevirt-hyperconverged'

$ cat deploy/hco.cr.yaml
---
apiVersion: hco.kubevirt.io/v1alpha1
kind: HyperConverged
metadata:
  name: foo
  namespace: kubevirt-hyperconverged

$ kubectl create -f deploy/hco.cr.yaml
The HyperConverged "foo" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"hco.kubevirt.io/v1alpha1", "kind":"HyperConverged", "metadata":map[string]interface {}{"creationTimestamp":"2020-03-04T17:40:49Z", "generation":1, "name":"foo", "namespace":"kubevirt-hyperconverged", "uid":"49a1bb72-5e3f-11ea-83f0-52fdfc072182"}}: validation failure list:
metadata.name in body should match 'kubevirt-hyperconverged'
```

Part 2, restricting the operators deployed namespace (deployed operator and all components into default namespace):

```
$ kubectl logs hyperconverged-cluster-operator-58cb96fc57-hs9qp
{"level":"info","ts":1583345316.2963555,"logger":"cmd","msg":"Go Version: go1.13.1"}
{"level":"info","ts":1583345316.2964191,"logger":"cmd","msg":"Go OS/Arch: linux/amd64"}
{"level":"info","ts":1583345316.2964365,"logger":"cmd","msg":"Version of operator-sdk: v0.15.1"}
{"level":"error","ts":1583345316.2965324,"logger":"cmd","msg":"Please re-deploy this operator into kubevirt-hyperconverged namespace","Expected.Namespace":"kubevirt-hyperconverged","Deployed.Namespace":"default","error":"Operator running in different namespace than expected","stacktrace":"github.com/go-logr/zapr.(*za
pLogger).Error\n\t/home/dzager/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nmain.main\n\t/home/dzager/go/src/github.com/kubevirt/hyperconverged-cluster-operator/cmd/hyperconverged-cluster-operator/main.go:95\nruntime.main\n\t/home/dzager/.local/share/go/src/runtime/proc.go:203"}
```

**Pull request merged**